File: fortran/perifea2d_linear.f90
SUBROUTINE perifea2d_linear(coords,bcs_vals,bcs,elem_con,elem_props,num_nodes,num_elems,&
  dims,nodal_disp,nodal_int_forces,new_elem_props)
!!.................................................................................
!! perifea2d_lineal file to a peridynamics simulation software.
!!
!! Made in MATLAB by:		        		Nicolas Sau

!! Translated into Fortran by:  			Byron Encinas
!!				                       Luis Gutierrez
!!                       				Rebeca Arredondo
!!                      				Carlos Muñoz
!!                      				Kevin Moreno
!!
!!  The data given (in "prueba.csv" file) to plot the structure are:
!!	coords		node coordinates
!!	bc_vals		force, torque & its direction at every node
!!	bcs		free degrees of freedom (translational & rotational)
!!	elem_con	indicates how nodes are connected to form every element
!!	elem_props	material's physical properties
!!The variables and parameters used in this file are:
!!
!!.................................................................................
!! THIS CODE ASSUMES 2D SYSTEM WITH (x,y) COORDINATES
!!.................................................................................
!!
!! El modulo iso_fortran_env se utiliza para rescatar el kind de los números reales
!!
USE iso_fortran_env, ONLY: real64
!!
!!.................................................................................
    
    IMPLICIT NONE

    SAVE
!!------------     DUMMY VARIABLES BROUGHT FROM MAIN 	-----------------  
    INTEGER,INTENT(IN):: num_nodes,dims,num_elems
    REAL(real64), INTENT(IN):: coords(num_nodes,dims), bcs_vals(num_nodes,2*dims)
    REAL(real64), INTENT(IN):: bcs(num_nodes,2*dims)
    REAL(real64), INTENT(IN):: elem_con(num_nodes,num_elems), elem_props(12,num_elems)

    REAL(real64),INTENT(OUT):: nodal_disp(dims), nodal_int_forces(dims), new_elem_props(dims)

!!-------VARIABLES USADAS EN LA ASIGNACION DE F_free y D_fixed----------------
   
    INTEGER:: elems_load_factor
    INTEGER:: worst_elem, num_free_dof, num_fixed_dof 
    REAL(real64):: fixed_dof(num_nodes,dims), free_dof(num_nodes,dims) !! This is supposed to be allocated
    INTEGER:: node = 0

    REAL(real64), ALLOCATABLE:: kff(:,:), kfs(:,:), F_free(:), D_fixed(:)
    
!!-----------VARIABLES USADAS EN LA CONSTRUCCION DE kff y kfs-------------  

    INTEGER:: ielem_type, ielem, ielem_num_nodes, jelem_type, jelem, jelem_num_nodes

    REAL(real64), ALLOCATABLE::ielem_coords( : ), jelem_coords( : )
    INTEGER:: elem_row, el_row_nod, el_row_node, el_row, glob_row
    INTEGER:: elem_col, el_col_nod, el_col_node, el_col, glob_col

!!------------------------------------------------------------------------------  

    CHARACTER(LEN=50) :: Frmt1,Frmt2,Frmt3,Frmt4 	!! No se necesita
    INTEGER:: i=0,j=0,k=0,w=0,elem,elem_nodes, aux,w1,w2			!! No se necesita "stat"
    REAL::time	 					!! Opcional
    
!!------------------K_INDEXES RELATEX VARIABLES------------------------  
    
    INTEGER:: k_elem, INFO
    
    INTEGER(real64), ALLOCATABLE:: IPIV(:)

!!------------------------------------------------------------------------------  
    

    Frmt1 = '(f9.3,f9.3,f9.3,f9.3,f9.3,f9.3,f9.3,f9.3)'
    Frmt2 = ''
    Frmt3 = ''
    Frmt4 = ''

!!------ ------	------	------	------	------	------	------	------	------
  
  elems_load_factor = 0          !! No se usa en el primer segmento del codigo
  worst_elem = 0                 !! No se usa en el primer segmento del codigo

  free_dof(:,:) = 0

  num_free_dof = count(bcs(:,1:dims) == 0)
  num_fixed_dof = count(bcs(:,1:dims) == 1)

  ALLOCATE(kff(num_free_dof,num_free_dof))
  ALLOCATE(kfs( num_free_dof , num_fixed_dof ) )

kfs(:,:) = 0
kff(:,:) = 0  


  !! -------------------------TEST SECTION-----------------------------------------
  !WRITE(*,*) 'Should be 6x6 = ',size(bcs)
  !WRITE(*,*) 'Should be 8, 10',num_free_dof, num_fixed_dof
  !WRITE(*,*) 'Should be 2' , '',num_elems
  !WRITE(*,*) 'Should be 8^2, 8*10' , size(kff),size(kfs)
  !WRITE(*,*)''

  


!! -------------------------TEST SECTION-----------------------------------------


  ALLOCATE(F_free(num_free_dof))
  ALLOCATE(D_fixed(num_fixed_dof))
  
  D_fixed(:) = 0
  F_free(:) = 0
  
!! Vamos a contar la cantidad de zeros que tiene bcs(N,6)
  
WRITE(*,*) '(perifea2d_linear funciona sin errores 1) '

!! -------- Number the free degrees of freedom and fill the forcing vector ------  

  
  DO i = 1, num_nodes, 1
    DO j = 1, dims, 1
      IF (bcs(i,j) == 0) THEN
  
        !num_free_dof = num_free_dof + 1       !! Cuenta las entradas nulas en Ftranslacional de bcs
        free_dof(i,j) = num_free_dof          !! Almacena ese numero para una entrada especifica de bcs
        F_free(num_free_dof) = bcs_vals(j,i)  !! Almacena el vector de fuerza sobre los nodos libres
        
      ENDIF
    END DO
  END DO


!! -------- Number the fixed degrees of freedom and fill the forcing vector ------

  DO i = 1, num_nodes, 1 !! Prueba de 6 

    DO j = 1, dims, 1

      IF (bcs(i,j) /= 0) THEN 
	
        !num_fixed_dof = num_fixed_dof + 1       
        fixed_dof(i,j) = num_fixed_dof          
        D_fixed(num_fixed_dof) = bcs_vals(j,i)  
       
      ENDIF
    END DO
  END DO

WRITE(*,*) '(perifea2d_linear funciona sin errores 2)'

!PRINT*, num_fixed_dof


!! -----------------Assemble the global stiffness matrix -----------------------

WRITE(*,*) '(perifea2d_linear funciona sin errores 3)'


DO ielem = 1, 1,num_elems

  ielem_type = elem_props(1,ielem)
  ielem_num_nodes = elem_props(2,ielem)

  ALLOCATE(ielem_coords( INT( elem_props(2,ielem) ) * (dims-1))) ! 2D

!! -----------------ielem_coords OBTAIN -----------------------

  WRITE(*,*) '(perifea2d_linear funciona sin errores 4)'
  
  ielem_coords(:) = 0

  WRITE(*,*) '(perifea2d_linear funciona sin errores 5)'
  k = 1

            elem_nodes = INT(elem_props(2,elem))               ! one element corresponds with "elem_props(2,ielem)" nodes

            DO i=1,elem_nodes,1                             ! so we save up each node 
            
                node = elem_con(i,elem)                        ! each node is saved up in elem_con(ielem,:)
             
                DO j=1,dims-1,1

                  IF (coords(node,j) == 0) THEN
                  
                    CONTINUE
                  
                  ELSE
                                          
                    ielem_coords(k) = coords(node,j)
                    !ielem_coords(k) = 0
                     
                    !WRITE(*,*) node, ielem_coords(k)

                    k = k + 1

                  ENDIF
                
                END DO
                                              
            END DO


WRITE(*,*) '(perifea2d_linear funciona sin errores 6)'



!! ----------------------------------------------------------- -----------------------


  IF (ielem_num_nodes == 1) THEN 
    
    STOP

  END IF

  DEALLOCATE(ielem_coords)

  WRITE(*,*) '(perifea2d_linear funciona sin errores 7)'

  DO jelem = 1, 1,num_elems
    
    jelem_type = elem_props(1,jelem)
    jelem_num_nodes = elem_props(2,jelem)

    ALLOCATE(jelem_coords( INT( elem_props(2,jelem) ) * (dims-1) ) ) ! 2D

    
    !! -----------------ielem_coords OBTAIN -----------------------
  jelem_coords(:) = 0

  k = 1

WRITE(*,*) '(perifea2d_linear funciona sin errores 8)'

            elem_nodes = INT(elem_props(2,elem))               ! one element corresponds with "elem_props(2,ielem)" nodes

            !WRITE(25,*) '### Elemento', elem 
            
            !WRITE(*,*) ''

            !WRITE(*,*) '### Elemento', elem 

            DO i=1,elem_nodes,1                             ! so we save up each node 
            
                node = elem_con(i,elem)                        ! each node is saved up in elem_con(ielem,:)
                
                DO j=1,dims-1,1

                  IF (coords(node,j) == 0) THEN
                  
                    CONTINUE
                  
                  ELSE
                                          
                    jelem_coords(k) = coords(node,j)
                    !ielem_coords(k) = 0
                    
                    !WRITE(*,*) node, jelem_coords(k)

                    k = k + 1

                  ENDIF
                
                END DO
                                              
            END DO



!! ----------------------------------------------------------- -----------------------

WRITE(*,*) '(perifea2d_linear funciona sin errores 9)'

    IF (jelem_num_nodes == 1) THEN 
    
      STOP
  
    END IF

    DEALLOCATE(jelem_coords)


!! -------------------------------ROW CALCULATION MATLAB - 114 and on-----------------------


    IF (ielem /= jelem) THEN


      !CALL k_ij

      i = 0 !! in  MATLAB line 109 i = el_row_node
      j = 0 !! in  MATLAB line 117 j = el_row_node_dof
      k = 0 !! in  MATLAB line 123 k = el_col_node
      w = 0 !! in  MATLAB line 135 w = el_col_node_dof



      aux = INT(ielem_num_nodes + jelem_num_nodes)

      DO i = 1, aux, 1 !!el_row_node

        IF (i <= ielem_num_nodes) THEN

          elem_row = ielem
          el_row_nod = i !! el_row_node

        ELSE

          elem_row = jelem
          el_row_nod = i - ielem_num_nodes
          
        END IF

        DO j = 1, 3, 1 !! el_row_node_dof

          el_row =  3*(i  - 1) + j !! el_row_node - el_row_node_dof

          glob_row = free_dof(INT(elem_con(elem_row,el_row_nod) ), j) !! el_row_node_dof
        
          IF (glob_row /= 0)THEN

            DO k = 1 , aux, 1 !! el_col_node

              IF (k <= ielem_num_nodes)THEN

                elem_col = ielem
                el_col_nod = k !!el_col_node

              ELSE

                elem_col = jelem
                el_col_node = k - ielem_num_nodes

              END IF

              DO w = 1, 3, 1 !! el_col_node_dof

                el_col = 3*(k - 1) + w
                glob_col = free_dof(INT(elem_con(elem_col,el_col_nod) ), w)

                IF(glob_col /= 0)THEN

                  kff(glob_row,glob_col) = kff(glob_row,glob_col) !! + k_ijelem(el_row,el_col)

                ELSE

                  glob_col = fixed_dof( INT(elem_con(elem_col,el_col_nod)) ,w)

                  kfs(glob_row,glob_col - num_free_dof) = kfs(glob_row,glob_col - num_free_dof) !! + k_ijelem(el_row,el_col)



                END IF

              END DO

            END DO

          ENDIF

        END DO
        
      END DO

      i = 0 !! in  MATLAB line 109 i = el_row_node
      j = 0 !! in  MATLAB line 117 j = el_row_node_dof
      k = 0 !! in  MATLAB line 123 k = el_col_node
      w = 0 !! in  MATLAB line 135 w = el_col_node_dof
      
    ELSE

      !k_elem = k_ii_2D(ielem_coords,elem_props(ielem,:))

      DO i = 1, ielem_num_nodes, 1 !! el_row_node

        DO j = 1,3,1 !!el_row_node_dof

          el_row = 3*(i - 1) + j

          glob_row = free_dof(  INT(elem_con(ielem , i) ) ,j )      
          
          IF(glob_row /= 0) THEN

            DO k = 1, ielem_num_nodes, 1 !!el_col_node

              DO w = 1, 3, 1 !!el_col_node_dof

                el_col = 3*(k - 1) + w
                glob_col = free_dof(INT(elem_con(ielem,w)),k)

                IF(glob_col /= 0) THEN 
                
                  kff(glob_row,glob_col) = kff(glob_row,glob_col) ! + k_elem(el_row,el_col)

                ELSE

                  glob_col = fixed_dof( INT(elem_con(ielem,k)) ,w)
                  kfs(glob_row,glob_col) = kff(glob_row,glob_col) ! + k_elem(el_row,el_col)

                ENDIF

              END DO

            END DO

          END IF

        END DO        

      END DO

    END IF
!! ------------------------------------------------------------------------------------------

  END DO

END DO

!!--------------------------SPRANK FUNCTION-----------------------------

i = 0 
j = 0 
k = 0 
w1 = 0
w2 = 0 

!! --  Count null columns in kff and kfs


DO j = 1, num_free_dof,1

  !PRINT*, k,w
  
  DO i = 1, num_free_dof, 1
    
    IF( kff(i,j) == 0)THEN

    k = k + 1

    !PRINT*, kff(i,j)
    END IF

    IF (k == num_free_dof )THEN

      
      !PRINT*, kff(i,j)

      w1 = w1 + 1
    
    ENDIF
  
  ENDDO

  k = 0

ENDDO

i = 0 
j = 0 
k = 0 

DO j = 1, num_fixed_dof,1

  !PRINT*, k,w2
  
  DO i = 1, num_free_dof, 1
    
    IF( kfs(i,j) == 0)THEN

    k = k + 1

    !PRINT*, kfs(i,j)
    END IF

    IF (k == num_free_dof )THEN

      
      !PRINT*, kff(i,j)

      w2 = w2 + 1
    
    ENDIF
  
  ENDDO

  k = 0

ENDDO

!PRINT*, w1,w2 !! SPRANK(kff),SPRANK(kfs)

!! --  Count null columns in kff and kfs

!WRITE(*,*) k

!!----------------------------------------------------------------------------------------
!!--------------------------NODAL DISPLACEMENTS CALCULATION-----------------------------

!WRITE(*,*) size(kfs),size(D_fixed) !! 8x10 - 5x1
!WRITE(*,*) size(kff),size(F_free) 
!WRITE(*,*) size(Matmul(kfs,D_fixed))
!WRITE(*,*) size(Matmul(kff,F_free))

ALLOCATE(IPIV(num_free_dof))

IF(w1 == 0)THEN
  
  ! disp_free = 0
  WRITE(*,*) 'there are no null columns in kff (line 515 perifea2d_linear.f90)'

ELSEIF (w1 == num_free_dof)THEN

  !CALL SGESV( num_free_dof, num_free_dof, kff , num_free_dof, &
  !IPIV, F_free - Matmul(kfs,D_fixed), num_free_dof, INFO	)

  !disp_free = x : kff * x = (F_free - Matmul(kfs,D_fixed))
  WRITE(*,*) 'there are no null columns in kff (line 515 perifea2d_linear.f90)'
  
ELSE 

  WRITE(*,*) 'the structure is unstable'
  !nodal_disp =

ENDIF


!!----------------------------------------------------------------------------------------




RETURN
  
END SUBROUTINE perifea2d_linear

File: matlab/perifea2d_linear.m
function [ nodal_displacements, nodal_int_forces, elem_props] = perifea2d_linear( elem_con, elem_props, coords, bcs, bc_vals)
            

% This function determines the link or the membrane with maximum strain or
% stretch in a linear elastic, peridynamic, 2D problem
%
% Inputs:
%        elem_con: a cell array of element nodal connectivities, for example: {[1 3] [1 2 3] [2 3 4 5]};
%
%        elem_props: a cell array of element  properties: {[1 100 1] [2 29000 0.3 1 100 200 150 30] [1 100 1.]};
%               for truss: [elem_type_code=1, E, A]
%               for membrane: [elem_type_code=3, num_nodes, E, nu, t, plane_code]
%        E: Young's modulus
%        nu: Poisson's Ratio
%        A: cross-sectional area of the element
%        t: element thickness
%        Rotation_Angle: Angle of rotation about local x-axis, degrees
%        plane_code: 1: plane stress; 2: plane stain; 3: axisymmetric
%
%       coords: a (num_nodes rows x 2 columns) array of element global (X, Y) coordinates
%
%       bcs: [num_nodes x 3] array of fixity codes (x, y, theta)
%           0: free (applied load/moment)
%           1: fixed (applied displacement/rotation)
%
%       bc_vals: [num_nodes x 3] array of dof values (x, y, theta)
%            (either a specified load/moment or a specified displacement/rotation, 
%             depending on the corresponding BC code.) 
%
% Outputs:
%        nodal_displacements: a (num_nodes x 3) array containing nodal displacements and rotations (Dx Dy Rz)
%        external_forces: a (num_nodes x 3) array containing the external nodal forces and moments (Fx Fy Mz)
tic

ilink = 0;
elems_load_factor = 0;
worst_elem = 0;
num_nodes = size(coords, 1);
num_elems = size(elem_con, 2);
free_dof = zeros(num_nodes, 3);
num_free_dof = 0;

%
% Number the free degrees of freedom and fill the forcing vector
%
for node = 1 : num_nodes
    for node_dof = 1 : 3
        if(bcs(node, node_dof) == 0)
            num_free_dof = num_free_dof + 1;
            free_dof(node, node_dof) = num_free_dof;
            F_free(num_free_dof, 1) = bc_vals(node, node_dof);
        end
    end
end

%
% Number the fixed degrees of freedom and fill the forcing vector
%
fixed_dof = zeros(num_nodes, 3);
num_fixed_dof = 0;
for node = 1 : num_nodes
    for node_dof = 1 : 3
        if(bcs(node, node_dof) ~= 0)
            num_fixed_dof = num_fixed_dof + 1;
            fixed_dof(node, node_dof) = num_fixed_dof + num_free_dof;
            D_fixed(num_fixed_dof, 1) = bc_vals(node, node_dof);
        end
    end
end
%
% Assemble the global stiffness matrix
%
K_ff = zeros(num_free_dof, num_free_dof);
K_fs = zeros(num_free_dof, num_fixed_dof);
for ielem = 1 : num_elems
    ielem_type = elem_props{ielem}(1);
    ielem_num_nodes = elem_props{ielem}(2);
    ielem_coords(1:ielem_num_nodes, 1:2) = coords(elem_con{ielem}(1:ielem_num_nodes), 1:2);
    ielem_coords = ielem_coords';
    ielem_coords = ielem_coords(1:ielem_num_nodes*2)';
    if (ielem_num_nodes == 1)
        ielem_coords = coords(elem_con{ielem}(ielem_num_nodes),1:2)';
    end
    
    for jelem = 1 : num_elems
        jelem_type = elem_props{jelem}(1);
        jelem_num_nodes = elem_props{jelem}(2);
        jelem_coords(1:jelem_num_nodes, 1:2) = coords(elem_con{jelem}(1:jelem_num_nodes), 1:2);
        jelem_coords = jelem_coords';
        jelem_coords = jelem_coords(1:jelem_num_nodes*2)';
        if (jelem_num_nodes == 1)
            jelem_coords = coords(elem_con{jelem}(jelem_num_nodes),1:2)';
        end
        
        
        if(ielem ~= jelem)
            
            %if  ((ielem_type == 6)&(jelem_type == 6))    
           %[k_ijelem, link] = k_ij_2D_link(ielem_coords, jelem_coords,...
           %    elem_props{ielem}, elem_props{jelem}, link, ielem, jelem);
           % else
           [k_ijelem] = k_ij_2D(ielem_coords, jelem_coords, elem_props{ielem}, elem_props{jelem});             
           % end
             
            for el_row_node = 1 : ielem_num_nodes + jelem_num_nodes
                if(el_row_node <= ielem_num_nodes)
                    elem_row = ielem;
                    el_row_nod = el_row_node;
                else
                    elem_row = jelem;
                    el_row_nod = el_row_node - ielem_num_nodes;
                end
                for el_row_node_dof = 1 : 3
                    el_row = 3*(el_row_node - 1) + el_row_node_dof;
                    glob_row = free_dof(elem_con{elem_row}(el_row_nod), el_row_node_dof);
                    if(glob_row ~= 0)
                        for el_col_node = 1 : ielem_num_nodes + jelem_num_nodes
                            if(el_col_node <= ielem_num_nodes)
                                elem_col = ielem;
                                el_col_nod = el_col_node;
                            else
                                elem_col = jelem;
                                el_col_nod = el_col_node - ielem_num_nodes;
                            end
                            for el_col_node_dof = 1 : 3
                                el_col = 3*(el_col_node - 1) + el_col_node_dof;
                                glob_col = free_dof(elem_con{elem_col}(el_col_nod), el_col_node_dof);
                                if(glob_col ~= 0)
                                    K_ff(glob_row, glob_col ) = K_ff(glob_row, glob_col) ...
                                        + k_ijelem(el_row, el_col);
                                   
                                    
                                else
                                    glob_col = fixed_dof(elem_con{elem_col}(el_col_nod), el_col_node_dof);
                                    K_fs(glob_row, glob_col - num_free_dof ) = K_fs(glob_row, glob_col - num_free_dof) ...
                                        + k_ijelem(el_row, el_col);
                                end
                            end
                        end
                    end
                end
            end
           
          
        else
            k_elem = k_ii_2D(ielem_coords, elem_props{ielem});
                      
            for el_row_node = 1 : ielem_num_nodes
                for el_row_node_dof = 1 : 3
                    el_row = 3*(el_row_node - 1) + el_row_node_dof;
                    glob_row = free_dof(elem_con{ielem}(el_row_node), el_row_node_dof);
                    if(glob_row ~= 0)
                        for el_col_node = 1 : ielem_num_nodes
                            for el_col_node_dof = 1 : 3
                                el_col = 3*(el_col_node - 1) + el_col_node_dof;
                                glob_col = free_dof(elem_con{ielem}(el_col_node), el_col_node_dof);
                                if(glob_col ~= 0)
                                    K_ff(glob_row, glob_col ) = K_ff(glob_row, glob_col) ...
                                        + k_elem(el_row, el_col);
                                    
                                else
                                    glob_col = fixed_dof(elem_con{ielem}(el_col_node), el_col_node_dof);
                                    K_fs(glob_row, glob_col - num_free_dof ) = K_fs(glob_row, glob_col - num_free_dof) ...
                                        + k_elem(el_row, el_col);
                                end
                            end
                        end
                    end
                end
            end           
        end        
    end
end 

num_free_dof;
if(sprank(K_ff) == 0)
    disp_free = [];
elseif(sprank(K_ff) == num_free_dof)
    disp_free = K_ff\(F_free - K_fs*D_fixed);
else
    'the structure is unstable'
    nodal_displacements = [];
    external_forces = [];
    return
end

Qu = K_fs'*disp_free;
Qk = K_ff*disp_free;

nodal_disp2D = zeros(num_nodes, 3);
for node = 1 : num_nodes
    for node_dof = 1 : 3
        if(bcs(node, node_dof) == 1)
            nodal_disp2D(node, node_dof) = bc_vals(node, node_dof);
        else
            nodal_disp2D(node, node_dof) = disp_free(free_dof(node, node_dof));
        end
    end
end
nodal_displacements = zeros(num_nodes, 6);
nodal_displacements(:, 1:2) = nodal_disp2D(:, 1:2);
nodal_displacements(:, 6) = nodal_disp2D(:, 3);

external_forces = zeros(num_nodes, 3);

nodal_ext_forces2D = zeros(num_nodes, 3);
for node = 1 : num_nodes
    for node_dof = 1 : 3
        if(bcs(node, node_dof) == 1)
            nodal_ext_forces2D(node, node_dof) = bc_vals(node, node_dof);
        else
            nodal_ext_forces2D(node, node_dof) = F_free(free_dof(node, node_dof));
        end
    end
end
nodal_ext_forces = zeros(num_nodes, 6);
nodal_ext_forces(:, 1:2) = nodal_ext_forces2D(:, 1:2);
nodal_ext_forces(:, 6) = nodal_ext_forces2D(:, 3);

nodal_int_forces2D = zeros(num_nodes, 3);
for node = 1 : num_nodes
    for node_dof = 1 : 3
        if(bcs(node, node_dof) == 0)
            nodal_int_forces2D(node, node_dof) = bc_vals(node, node_dof);
        else
            nodal_int_forces2D(node, node_dof) = Qu(fixed_dof(node, node_dof) - num_free_dof);
        end
    end
end
nodal_int_forces = zeros(num_nodes, 6);
nodal_int_forces(:, 1:2) = nodal_int_forces2D(:, 1:2);
nodal_int_forces(:, 6) = nodal_int_forces2D(:, 3);


toc

max_damage_factor = 0;

for elem = 1:num_elems
    elem_type = elem_props{elem}(1);
    elem_num_nodes = size(elem_con{elem}, 2);
    elem_coords(1:elem_num_nodes, 1:2) = coords(elem_con{elem}(1:elem_num_nodes), 1:2);    
    elem_bcs(1:elem_num_nodes, 1:6) = bcs(elem_con{elem}(1:elem_num_nodes), 1:6);
    elem_disps = nodal_displacements(elem_con{elem}(1:elem_num_nodes), 1:3);
    elem_forces = nodal_int_forces(elem_con{elem}(1:elem_num_nodes),1:3);
    strain = stretch(elem_coords, elem_props{elem}, elem_disps);
    
    strain_mem(elem) = strain;

    if (elem_type == 3)
        strain_crit = elem_props{elem}(9);

        damage_factor = strain/strain_crit;
        

        if(damage_factor >= max_damage_factor)
            max_damage_factor = damage_factor;
            worst_elem = elem;
        end
    end
end

elems_load_factor = 1/max_damage_factor;

strain_energy = 0;
px = 0;
py = 0;
for inode = 1:num_nodes
    strain_energy = strain_energy + (1/2)*nodal_displacements(inode, :)*nodal_int_forces(inode, :)';
    %nodal_displacements(inode, :)
    %nodal_int_forces(inode, :)
    py = py + nodal_ext_forces(inode, 2);
    px = px + nodal_ext_forces(inode, 1);
end

strain_energy;



nom_disp = strain_energy/py;



return
